### PR TITLE
configure new plugin: ml-commons

### PR DESCRIPTION
**Pull request by ylwu-amzn**
Signed-off-by: Yaliang Wu <ylwu@amazon.com>

### Description
Configure new plugin: ml-commons.
 
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ylwu-amzn** [comment]
[check (manifests/1.3.0/opensearch-1.3.0.yml)](https://github.com/opensearch-project/opensearch-build/runs/5413075885?check_suite_focus=true) failed
```
FAILURE: Build failed with an exception.

* Where:
Build file '/tmp/tmp4eauaas5/k-NN/build.gradle' line: 51

* What went wrong:
A problem occurred evaluating root project 'opensearch-knn'.
> Failed to apply plugin class 'org.opensearch.gradle.info.GlobalBuildInfoPlugin'.
   > Could not create plugin of type 'GlobalBuildInfoPlugin'.
      > Could not generate a decorated class for type GlobalBuildInfoPlugin.
         > org/gradle/jvm/toolchain/JavaInstallation

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org/

BUILD FAILED in 11s
2022-03-03 20:29:03 ERROR    CI Manifest check failed
Traceback (most recent call last):
  File "./src/run_ci.py", line 25, in <module>
    sys.exit(main())
  File "./src/run_ci.py", line 21, in main
    CiManifests.from_file(args.manifest, args).check()
  File "/home/runner/work/opensearch-build/opensearch-build/src/ci_workflow/ci_manifest.py", line 21, in check
    self.__check__()
  File "/home/runner/work/opensearch-build/opensearch-build/src/ci_workflow/ci_input_manifest.py", line 36, in __check__
    ci_check_list.check()
  File "/home/runner/work/opensearch-build/opensearch-build/src/ci_workflow/ci_check_list_source.py", line 43, in check
    instance = klass(self.component, self.git_repo, self.target, check.args)
  File "/home/runner/work/opensearch-build/opensearch-build/src/ci_workflow/ci_check_gradle_properties.py", line 19, in __init__
    self.properties = self.__get_properties()
  File "/home/runner/work/opensearch-build/opensearch-build/src/ci_workflow/ci_check_gradle_properties.py", line 30, in __get_properties
    return PropertiesFile(self.git_repo.output(cmd))
  File "/home/runner/work/opensearch-build/opensearch-build/src/git/git_repository.py", line 78, in output
    return subprocess.check_output(command, cwd=cwd, shell=True).decode().strip()
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/subprocess.py", line 411, in check_output
    **kwargs).stdout
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/subprocess.py", line 512, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command './gradlew properties -Dopensearch.version=1.3.0-SNAPSHOT -Dbuild.snapshot=true' returned non-zero exit status 1.
Error: Process completed with exit code 1.
```

**Comment by gaiksaya** [comment]
Hi @opensearch-project/k-nn , can someone take a look at the above error?
The build for k-nn is failing

**Comment by gaiksaya** [comment]
Hi @ylwu-amzn can you rebase with main? The issue should be fixed now.
Thanks!

**Comment by ylwu-amzn** [comment]
> Hi @ylwu-amzn can you rebase with main? The issue should be fixed now. See #1704 Thanks!

Thanks @gaiksaya , already rebased, waiting for check workflow

**Comment by ylwu-amzn** [comment]
https://github.com/opensearch-project/opensearch-build/blob/main/ONBOARDING.md#onboard-to-build-workflow
>4. Execute ./build.sh to ensure your component builds and all artifacts are correctly placed into ./artifacts/ with correct output names.

`./build.sh manifests/1.3.0/opensearch-1.3.0.yml` failed to build k-NN. Not sure why, @gaiksaya @jmazanec15 can you help confirm if k-NN can pass `./build.sh` now? Is this some known issue?

After removing k-NN from `manifests/1.3.0/opensearch-1.3.0.yml` , `./build.sh manifests/1.3.0/opensearch-1.3.0.yml` ran successfully. Can see `opensearch-ml-1.3.0.0.zip` in `builds/opensearch/plugins`
```
% ll builds/opensearch/plugins
-rw-r--r-- 1 ylwu amazon  11M Mar  4 09:13 opensearch-alerting-1.3.0.0.zip
-rw-r--r-- 1 ylwu amazon  13M Mar  4 09:21 opensearch-anomaly-detection-1.3.0.0.zip
-rw-r--r-- 1 ylwu amazon 503K Mar  4 09:13 opensearch-asynchronous-search-1.3.0.0.zip
-rw-r--r-- 1 ylwu amazon 5.0M Mar  4 09:22 opensearch-cross-cluster-replication-1.3.0.0.zip
-rw-r--r-- 1 ylwu amazon 8.1M Mar  4 09:14 opensearch-index-management-1.3.0.0.zip
-rw-r--r-- 1 ylwu amazon 1.2M Mar  4 09:12 opensearch-job-scheduler-1.3.0.0.zip
-rw-r--r-- 1 ylwu amazon  17M Mar  4 09:25 opensearch-ml-1.3.0.0.zip
-rw-r--r-- 1 ylwu amazon 6.3M Mar  4 09:24 opensearch-observability-1.3.0.0.zip
-rw-r--r-- 1 ylwu amazon  59M Mar  4 09:20 opensearch-performance-analyzer-1.3.0.0.zip
-rw-r--r-- 1 ylwu amazon 6.8M Mar  4 09:23 opensearch-reports-scheduler-1.3.0.0.zip
-rw-r--r-- 1 ylwu amazon  39M Mar  4 09:17 opensearch-security-1.3.0.0.zip
-rw-r--r-- 1 ylwu amazon  18M Mar  4 09:23 opensearch-sql-1.3.0.0.zip

```

**Comment by ylwu-amzn** [comment]
Run `./assemble.sh builds/opensearch/manifest.yml` successfully after `./build.sh manifests/1.3.0/opensearch-1.3.0.yml` (with k-NN removed from `manifests/1.3.0/opensearch-1.3.0.yml`).

```
2022-03-04 09:44:18 INFO     Installed plugins: ['opensearch-alerting', 'opensearch-anomaly-detection', 'opensearch-security', 'opensearch-index-management', 'opensearch-sql', 'opensearch-observability', 'opensearch-cross-cluster-replication', 'opensearch-ml', 'opensearch-performance-analyzer', 'opensearch-job-scheduler', 'opensearch-asynchronous-search', 'opensearch-reports-scheduler']
2022-03-04 09:44:58 INFO     Published /local/home/ylwu/code/os/opensearch-build/dist/opensearch/opensearch-1.3.0-linux-x64.tar.gz.
2022-03-04 09:44:58 INFO     Done.
```

**Review by gaiksaya — APPROVED** [review]

**Comment by gaiksaya** [comment]
Hi @dblock , do we need to add build script for ml here: https://github.com/opensearch-project/opensearch-build/tree/main/scripts/components?

**Comment by tianleh** [comment]
> Hi @dblock , do we need to add build script for ml here: https://github.com/opensearch-project/opensearch-build/tree/main/scripts/components?

This is the script finder logic https://github.com/opensearch-project/opensearch-build/blob/main/src/paths/script_finder.py#L23

Since it is not added under dir above and also no repo level build script, it will use the default one https://github.com/opensearch-project/opensearch-build/blob/main/scripts/default/opensearch/build.sh

As long as @ylwu-amzn is fine with the default script, there shall be no concern.

**Comment by ylwu-amzn** [comment]
> Hi @dblock , do we need to add build script for ml here: https://github.com/opensearch-project/opensearch-build/tree/main/scripts/components?

@peterzhuamazon helped explained how maven publishing works and I have published a PR https://github.com/opensearch-project/ml-commons/pull/165 to add `build.sh` to `ml-commons`. 

Tested locally `./build.sh manifests/1.3.0/opensearch-1.3.0.yml`, then can see `opensearch-ml-client`.  @peterzhuamazon can you help confirm if that means all done? Should I change any other place?
```
% ls maven/org/opensearch/opensearch-ml-client
1.3.0.0  maven-metadata.xml  maven-metadata.xml.md5  maven-metadata.xml.sha1  maven-metadata.xml.sha256  maven-metadata.xml.sha512
```
And I also ran `./assemble.sh builds/opensearch/manifest.yml`, install the built tarball and start local cluster on desktop, it works well, both SQL and MLCommons.

**Review by tianleh — APPROVED** [review]

File: manifests/1.3.0/opensearch-1.3.0-test.yml [hunk 1]
```diff
@@ -88,4 +88,14 @@ components:
     bwc-test:
       test-configs:
         - without-security
+
```
Comment (peterzhuamazon) [inline manifests/1.3.0/opensearch-1.3.0-test.yml:91]: Remove newline

File: manifests/1.3.0/opensearch-1.3.0.yml [hunk 2]
```diff
@@ -26,6 +26,12 @@ components:
     checks:
       - gradle:properties:version
       - gradle:dependencies:opensearch.version
+  - name: ml-commons
+    repository: https://github.com/opensearch-project/ml-commons.git
+    ref: main
+    checks:
+      - gradle:properties:version
+      - gradle:dependencies:opensearch.version: opensearch-ml-plugin
```
Comment (peterzhuamazon) [inline manifests/1.3.0/opensearch-1.3.0.yml:34]: I dont think you need to add `: opensearch-ml-plugin` after.

**Review by peterzhuamazon — CHANGES_REQUESTED** [review]
A bit concerned about the naming.
The repo is `ml-commons`, then there is name `opensearch-ml-plugin`, and this PR https://github.com/opensearch-project/ml-commons/pull/165 suggest the maven name being `opensearch-ml-client`.

Can we have a consistent naming convention here?

Thanks.

File: manifests/1.3.0/opensearch-1.3.0.yml (hunk 2 above)
Comment (peterzhuamazon) [reply manifests/1.3.0/opensearch-1.3.0.yml:34]: Got it.

So is the naming `opensearch-ml-plugin` or `opensearch-ml-client`?

File: manifests/1.3.0/opensearch-1.3.0-test.yml (hunk 1 above)
Comment (peterzhuamazon) [reply manifests/1.3.0/opensearch-1.3.0-test.yml:91]: Seems like I mixed up the test and non-test file.
Sorry.

File: manifests/1.3.0/opensearch-1.3.0.yml (hunk 2 above)
Comment (peterzhuamazon) [reply manifests/1.3.0/opensearch-1.3.0.yml:34]: Just talked to @ylwu-amzn as their project is actually a combination of 4 subproject, where opensearch-ml-plugin is the top level project with all dependency trees. And opensearch-ml-client is one of the subproject that required to be on maven.

Thanks.

**Review by peterzhuamazon — APPROVED** [review]

**Pull request merged**